src/Servers/Kestrel/Transport.NamedPipes/src/Internal/NamedPipeTransportFactory.cs

```diff
@@ -45,19 +45,22 @@ public ValueTask<IConnectionListener> BindAsync(EndPoint endpoint, CancellationT
             throw new NotSupportedException($@"Server name '{namedPipeEndPoint.ServerName}' is invalid. The server name must be ""{LocalComputerServerName}"".");
         }
 
-        // Creating a named pipe server with an name isn't exclusive. Create a mutex with the pipe name to prevent multiple endpoints
-        // accidently sharing the same pipe name. Will detect across Kestrel processes.
-        // Note that this doesn't prevent other applications from using the pipe name.
-        var mutexName = "Kestrel-NamedPipe-" + namedPipeEndPoint.PipeName;
-        var mutex = new Mutex(false, mutexName, out var createdNew);
-        if (!createdNew)
+        var listener = new NamedPipeConnectionListener(namedPipeEndPoint, _options, _loggerFactory, _objectPoolProvider);
+
+        // Start the listener and create NamedPipeServerStream instances immediately.
+        // The first server stream is created with the FirstPipeInstance flag. The FirstPipeInstance flag ensures
+        // that no other apps have a running pipe with the configured name. This check is important because:
+        // 1. Some settings, such as ACL, are chosen by the first pipe to start with a given name.
+        // 2. It's easy to run two Kestrel app instances. Want to avoid two apps listening on the same pipe and creating confusion.
+        //    The second launched app instance should immediately fail with an error message, just like port binding conflicts.
+        try
         {
-            mutex.Dispose();
-            throw new AddressInUseException($"Named pipe '{namedPipeEndPoint.PipeName}' is already in use by Kestrel.");
+            listener.Start();
+        }
+        catch (UnauthorizedAccessException ex)
+        {
+            throw new AddressInUseException($"Named pipe '{namedPipeEndPoint.PipeName}' is already in use by Kestrel.", ex);
```

Comment:
nit: in theory some other app could generate exactly the same named pipe name and start it first, so there is no guarantee that it's Kestrel

```suggestion
            throw new AddressInUseException($"Named pipe '{namedPipeEndPoint.PipeName}' is already in use (most likely by Kestrel).", ex);
```

Reply:
Good point. I'll remove mentioning Kestrel altogether.